### PR TITLE
ISPN-8245 Make Hibernate Cache provider work with Hibernate 5.1

### PR DESCRIPTION
**Pull request by galderz**
We need to get this into Infinispan 9.1.1 so that Wildfly can consume it.

### CODE REVIEW

**Comment by galderz** [comment]
Test failure has nothing to do with PR.

**Comment by tristantarrant** [comment]
@pferraro we need you to validate this.

File: hibernate-cache/src/test/java/org/infinispan/test/hibernate/cache/access/PutFromLoadValidatorUnitTest.java [hunk 1]
```diff
@@ -273,7 +273,7 @@ private void multipleRegistrationtest(final boolean transactional) throws Except
 				if (transactional) {
 					tm.begin();
 				}
-				SharedSessionContractImplementor session = mock (SharedSessionContractImplementor.class);
+				SessionImplementor session = mock (SessionImplementor.class);
```
Comment (rvansa) [inline hibernate-cache/src/test/java/org/infinispan/test/hibernate/cache/access/PutFromLoadValidatorUnitTest.java]: Remove space between `mock` and `(` please (style should be corrected on modified lines).

File: hibernate-cache/src/test/java/org/infinispan/test/hibernate/cache/functional/AbstractNonInvalidationTest.java [hunk 2]
```diff
@@ -128,11 +130,11 @@ public void cleanup() throws Exception {
                awaitOrThrow(preFlushLatch);
             }
             s.flush();
-         } catch (OptimisticLockException e) {
+         } catch (StaleStateException e) {
```
Comment (rvansa) [inline hibernate-cache/src/test/java/org/infinispan/test/hibernate/cache/functional/AbstractNonInvalidationTest.java]: please use `OptimisticLockException | StaleStateException e`

File: hibernate-cache/src/test/java/org/infinispan/test/hibernate/cache/tm/JBossStandaloneJtaExampleTest.java [hunk 3]
```diff
@@ -108,12 +108,10 @@ public void testPersistAndLoadUnderJta() throws Exception {
 			ut.begin();
 			try {
 				Session session = sessionFactory.openSession();
-				assertTrue(session.getTransaction().isActive());
+            session.getTransaction().begin();
```
Comment (rvansa) [inline hibernate-cache/src/test/java/org/infinispan/test/hibernate/cache/tm/JBossStandaloneJtaExampleTest.java]: You should not work with Hibernate transactions explicitly - everything should be controlled from the JTA transaction object/through `j.t.TransactionManager`

File: hibernate-cache/src/test/java/org/infinispan/test/hibernate/cache/tm/JBossStandaloneJtaExampleTest.java [hunk 4]
```diff
@@ -108,12 +108,10 @@ public void testPersistAndLoadUnderJta() throws Exception {
...
 				item = new Item("anItem", "An item owned by someone");
 				session.persist(item);
-				// IMO the flush should not be necessary, but session.close() does not flush
-				// and the item is not persisted.
-				session.flush();
+            session.getTransaction().commit();
```
Comment (rvansa) [inline hibernate-cache/src/test/java/org/infinispan/test/hibernate/cache/tm/JBossStandaloneJtaExampleTest.java]: The idea is that you close the session before JTA transaction is committed. Besides, this wouldn't use the correct transaction handling pattern.

File: hibernate-cache/src/test/java/org/infinispan/test/hibernate/cache/functional/AbstractNonInvalidationTest.java (hunk 2 above)
Comment (galderz) [reply hibernate-cache/src/test/java/org/infinispan/test/hibernate/cache/functional/AbstractNonInvalidationTest.java]: I'm not sure about this. 

`StateStateException` is what's expected in Hibernate 5.1, and `OptimisticLockException` in Hibernate 5.2. To use `OptimisticLockException | StaleStateException e` here, you'd be implying that in Hibernate 5.1, either can be thrown, which is not strictly true...

Your suggestion is convenient in the sense that it allows you easily deal with either Hibernate version though. Not fully convinced by it though.

**Comment by galderz** [comment]
@rvansa I've sent a commit fixing the comments agreed upon.

File: hibernate-cache/src/test/java/org/infinispan/test/hibernate/cache/tm/JBossStandaloneJtaExampleTest.java [hunk 5]
```diff
@@ -108,7 +108,6 @@ public void testPersistAndLoadUnderJta() throws Exception {
 			ut.begin();
 			try {
 				Session session = sessionFactory.openSession();
-				assertTrue(session.getTransaction().isActive());
```
Comment (rvansa) [inline hibernate-cache/src/test/java/org/infinispan/test/hibernate/cache/tm/JBossStandaloneJtaExampleTest.java]: `assertEquals(TransactionStatus.ACTIVE, session.getTransaction().getStatus());`

**Comment by galderz** [comment]
@rvansa I've fixed the latest comment. If you're happy, I'll squash and update branch.

**Comment by rvansa** [comment]
Sure, do so.

**Comment by galderz** [comment]
@rvansa Rebased

**Comment by rvansa** [comment]
There were no failures in the build before squash, integrated. Thanks.

**Pull request merged**